### PR TITLE
Making reporter Cukes compatible

### LINKED ISSUE #21: [Feature request] Add support for cucumber
**Describe the bug**
When using cucumber as a framework, video recording fails to work. No videos are produced. Cluttered and confusing raw output captured.

**Log**
https://gist.github.com/tangentially/aaf0c0d20364af9cc3279ba1ae120e9d

**To Reproduce**
Create a simple cucumber scenario and run.

**Expected behavior**
Videos recorded, one per scenario.

**Environment (please complete the following information):**
 - **wdio-video-reporter version:** 1.4.4
 - **WebdriverIO version:** 5.12.0
 - **Mode:** WDIO Testrunner
 - **If WDIO Testrunner, running sync/async:** sync
 - **Node.js version:** 10.16.0
 - **NPM version:** 6.10.3
 - **Browser name and version:** Chrome Version 74.0.3729.169 (Official Build) (64-bit)
 - **Platform name and version:** MacOS

**Desktop (please complete the following information):**
 - OS: MacOS
 - Browser Chrome 
 - Version 74

**Additional context**
From wdio-cucumberjs-json-reporter

The order of running of the on* is the following:
onRunnerStart
onSuiteStart (feature)
onSuiteStart (scenario)
onHookStart
onHookEnd
onTestStart
onBeforeCommand
onAfterCommand
onTestPass
onHookStart
onHookEnd
onSuiteEnd (scenario is done)
onSuiteEnd (feature is done)
onRunnerEnd

Please note: onTestEnd() is never called, and that onTestStart() is called once per step. (This leads to a _lot_ of single screenshot directories.

ls rawSeleniumVideoGrabs/
CES-Demo--HAOD-1325-Demo-scenario---subscribe-to-each-package-start-guidance-and-add-waypoints--Given-I-am-on-the-Login-page--CHROME--08-14-2019--11-40-58-918
CES-Demo--HAOD-1325-Demo-scenario---subscribe-to-each-package-start-guidance-and-add-waypoints--Given-I-am-on-the-Login-page--Then-the-password-text-input-is-not-empty--When-I-click-the-Login-button-on-the-Login-page--CHROME--08-14-2019--11-41-04-280
CES-Demo--HAOD-1325-Demo-scenario---subscribe-to-each-package-start-guidance-and-add-waypoints--Given-I-am-on-the-Login-page--When-I-type-drivertest3-username--CHROME--08-14-2019--11-41-00-152
CES-Demo--HAOD-1325-Demo-scenario---subscribe-to-each-package-start-guidance-and-add-waypoints--Given-I-am-on-the-Login-page--When-I-type-drivertest3-username--Then-the-username-text-input-has-value-drivertest3--CHROME--08-14-2019--11-41-01-732

### PR DESCRIPTION
**Pull request by jonn-set**
fixes #21 
- Works with Cucumber JS v6 (not tested on other versions)
- Includes tests for flows written
- Works with Allure report if `useCucumberStepReporter` option is set to `true`
- One problem I noticed is when the video file is more than 256 KB, it will only copy 256 KB to the allure report folder. But this is intermittent.
- Changed the Travis file to YML file, sorry :)
- Travis build is passing [here]([https://travis-ci.com/jonn-set/wdio-video-reporter/builds/138435811)

### CODE REVIEW

**Comment by jonn-set** [comment]
Hi, please can someone review and give feedback, please

**Comment by NavinSeth** [comment]
This is great and I can use this to for Cucumber. Can we please get a priority on this :).

File: .travis.yml [hunk 1]
```diff
@@ -1,10 +1,11 @@
...
-    "sh -e /etc/init.d/xvfb start",
-    "yarn"
-  ],
-  "script": "npm run ci"
-}
+node_js:
+  - '8.12.0'
+
+services:
+  - xvfb
```
Comment (presidenten) [inline .travis.yml]: Much nicer with yaml! 
Thanks!

**Comment by presidenten** [comment]
I have started to have a look at the code. But I have now way of verifying this since I dont know how to setup a cucumber project.  The previous project someone made for me to check cucumber with wdio-video-reporter doesnt want to run for some reason.

Would you mind forking and converting the template, so that it uses cucumber?
I created a brancher here: https://github.com/presidenten/wdio-template/tree/cucumber, would be nice to have a PR there so I can reference it when people wants to file issues in the future.

**Comment by jonn-set** [comment]
> I have started to have a look at the code. But I have now way of verifying this since I dont know how to setup a cucumber project. The previous project someone made for me to check cucumber with wdio-video-reporter doesnt want to run for some reason.
> 
> Would you mind forking and converting the template, so that it uses cucumber?
> I created a brancher here: https://github.com/presidenten/wdio-template/tree/cucumber, would be nice to have a PR there so I can reference it when people wants to file issues in the future.

Thanks for that, I have been testing it on my machine for a project I am working on by using `npm link` and things work as expected barring one problem which I have fixed but not pushed yet. I will be pushing it during the weekend or early next week as I need to write some tests for the new changes I have made, but you can review the rest or anyone interested in these changes can pull and link it to your project using `npm link` and run it.

**Comment by presidenten** [comment]
Oups, made a mistake while extracting framework code to separate files. Hold on, fix is coming.

**Comment by presidenten** [comment]
There we go. Should work as intended now :-)

Please insert your changes in `./src/frameworks/cucumber.js`

**Comment by presidenten** [comment]
@jonn-set 
Managed to get the cucumber example running after all, I put it in a branch here:
https://github.com/presidenten/wdio-template/tree/cucumber

So there is no need for you to fix an example for me.

**Comment by presidenten** [comment]
@jonn-set 
I extended the small demo to also run a few cucumber features, all though I cant in any way make it continue running the feature that should pass, after the first feature fails (any ideas?)

Any how, after you build normally, just head in to the `demo` directory, run `yarn`, and then `yarn e2e-cucumber`. The videos seem to pop up like they should, but if you run `yarn report` it seems like the yarn report shows the failed feature as broken instead of failed.

Any idea how to fix this?

I think we are pretty close though. Really nice job with the implementation.

**Comment by presidenten** [comment]
@jonn-set 
Sorry, lots of spam now. Updated the readme as well.
Please have look at the configuration instructions.

...and it would be great if you could have a look at cucumber demo as well
- `./demo/wdio-cucumber.conf.js`
- `./demo/cucumber-scenarios`

Now Im signing of for a while until you have had time to process the changes :-)

**Comment by jonn-set** [comment]
Thanks for looking at this. Sorry, weekends are a bit hectic for me sometimes, I will have a look at this when I get some time, hopefully at least by Monday.

Thanks

**Comment by jonn-set** [comment]
Sorry, accidentally hit the wrong button, hence reopened.

**Comment by presidenten** [comment]
Dont worry about it, have a look when you have time. We stress enough at work, lets not stress over hobby stuff as well :-)

**Comment by jonn-set** [comment]
Firstly, great work refactoring the whole thing so quickly. Please find answers to your questions below.

> @jonn-set
> I extended the small demo to also run a few cucumber features, all though I cant in any way make it continue running the feature that should pass, after the first feature fails (any ideas?)
> 
Yes, its because `bail `is set to `1` in the config file, should be `0`, will fix and push this.
> Any how, after you build normally, just head in to the `demo` directory, run `yarn`, and then `yarn e2e-cucumber`. The videos seem to pop up like they should, but if you run `yarn report` it seems like the yarn report shows the failed feature as broken instead of failed.
> 
> Any idea how to fix this?
> 
Yes, this is because of allure limitation, looks like it does not categorise node assert fails as a test failure. I normally use chai assertion and they appear correctly as failed, I wouldn't worry about it, but if you want I can add chai as a dev dependency to demo and edit the steps so that it appears correctly on the report.
> I think we are pretty close though. Really nice job with the implementation.

I have pushed my changes, have a go at it and let me know what you think.

Thanks

**Comment by jonn-set** [comment]
I have added a fix for cucumber where the `onTestSkip `function would delete the folder where scenario failed with a few steps still remaining to execute. I have run all the possible tests and they all seem to work as expected. @presidenten if you could check for other frameworks and also do a review of my changes and are happy, then looks like its good to release.

**Comment by presidenten** [comment]
@jonn-set I dived into the 256KB problem you mentioned in your first post and have rewritten the "finished"-handler to be able to handle bigger videos. Silly mistake from my side, but I also found the `isSynchronised` reporter-getter that allows me to use async functions instead of the ugly wait.

Im doing some refactoring now, then Ill let you test it. 
If all looks good, we'll merge to master and make a release.

Also, did you check the changes in readme? Does the configurations section for cucumber look ok?

**Comment by jonn-set** [comment]
> @jonn-set I dived into the 256KB problem you mentioned in your first post and have rewritten the "finished"-handler to be able to handle bigger videos. Silly mistake from my side, but I also found the `isSynchronised` reporter-getter that allows me to use async functions instead of the ugly wait.
>
@presidenten Thanks for looking into the 256KB issue, there seem to be another issue where videos gets copied with a size of a few bytes to the `allure-raw` folder, The files are generated correctly in the `rawSeleniumVideoGrabs` folder. I saw it happening for the 2nd browser that it was running the tests in parallel, so for me it was chrome most of the times, the firefox video was fine. This is intermittent and does not seem to happen on all runs. Could be an issue with Windows :)

> Im doing some refactoring now, then Ill let you test it.
>
Sure, will do.

> If all looks good, we'll merge to master and make a release.
>
That would be great, thanks. 

> Also, did you check the changes in readme? Does the configurations section for cucumber look ok?
>
I just looked at the readme and I dont think we need to tell users how to configure cucumber in wdio.conf file, I am sure people using cucumber will already know how to configure it, there a lot of boilerplate code for cucumber already on WDIO website. I have done some minor rephrasing of what you had in there for using it with allure, I will push it, you can proof read it and then change it accordingly.

**Comment by presidenten** [comment]
@jonn-set Nice update to readme. This is much better.

These new commits contains a bigger rewrite that should fix alot of problems with allure videos not being updated properly, as well as some more refactoring.
Now >256kb videos work as intended
image

With the updates to the feature files, I noticed that we get more folders in rawSeleniumVideoGrabs than actual videos. Is 
image

Are we missing something?

**Comment by jonn-set** [comment]
> @jonn-set Nice update to readme. This is much better.
> 
> These new commits contains a bigger rewrite that should fix alot of problems with allure videos not being updated properly, as well as some more refactoring.
> Now >256kb videos should hopefully work as intended...
> image
> 
> With the updates to the feature files, I noticed that we get more folders in rawSeleniumVideoGrabs than actual videos. Is
> image
> 
> Are we missing something?

The last two folders are of passing tests, it does not generate videos for passing tests if `saveAllVideos` is set to `false` :-)

**Comment by presidenten** [comment]
Lol! Ofcourse hahhahahaha. 
Stayed up to late last night to try to fix the rewrite and got a bit mushy. :-)

You should be able to test the >250kb with this step btw
```javascript
Given(/^I open Basic Examples tab$/, () => {
  for(let i=0; i<5; i++) {
    browser.url('http://www.seleniumeasy.com/test/');
    $('#btn_basic_example').click();
    browser.pause(300); // avoid animation effect
    $('.list-group-item[href*="basic-first-form-demo"]').click();
    $('#gettotal input').setValue('Presidenten');
  }

  $('#btn_basic_example').click();
  browser.pause(300); // avoid animation effect
});
```

**Comment by jonn-set** [comment]
The problem with the allure file copying is still happening, can you check if you see the same issue intermittently, please?

image

**Comment by presidenten** [comment]
I have run the tests in the demo dir a bunch of times now, 3 with selenium-docker and 3 with standalone-service, and 3 times on my much faster work computer (docker selenium).
But I cant seem to get the issue to happen. 

Its very strange that it happens, since I dont move on until all ffmpeg-processes have finished. Maybe the videoRenderTimeout is triggered?

Could you set loglevel to debug and send me all wdio logs on a run where it happens?
Also, please list node version so I can try to replicate it.

**Comment by presidenten** [comment]
Since I was on my work computer I ran the cucumber demo a few times on our dev-grid as well.
All desktop browsers are working:
- chrome ✓
- firefox ✓
- safari ✓
- ie11 ✓
- edge ✓

Our devices are acting up, but they havn't gotten any love for a while, and cant check on them since Im at home. So cant verify that it works.

**Comment by jonn-set** [comment]
> I have run the tests in the demo dir a bunch of times now, 3 with selenium-docker and 3 with standalone-service, and 3 times on my much faster work computer (docker selenium).
> But I cant seem to get the issue to happen.
> 
> Its very strange that it happens, since I dont move on until all ffmpeg-processes have finished. Maybe the videoRenderTimeout is triggered?
> 
> Could you set loglevel to debug and send me all wdio logs on a run where it happens?
> Also, please list node version so I can try to replicate it.

Its intermittent and even I cant get it reproduced every single time. The node version I am using is 10.13

**Comment by jonn-set** [comment]
> Its intermittent and even I cant get it reproduced every single time. The node version I am using is 10.13

Tried in Linux with node v10 and it seems to be working alright, might be a problem with windows then. I wouldn't worry about it until other users also complain about it once released.

**Comment by presidenten** [comment]
Thanks for the env-input. 

I managed to replicate the issue consistently on a Win10 virtual machine.

No clue why windows in behaving differently, but I have managed to make a fix that seem to make it work consistently. Just have to update tests.

Will write them later tonight.

**Comment by presidenten** [comment]
Could you check if this solves the 256k and 1k issue?

I noticed a couple of things that needed fixing.
- Sometimes (but not always), but especially on jasmine timeouts, there is a severe error and wdio skips the onRunnerEnd callback
- Sometimes on win, even though the child_process has finished, the video file is not done being saved when it was time to move the video files to allure.

So I moved the allure code to process.on('exit') to be a bit extra resilient, and I also wrote a newer better piece of code to check that videos were done being written to disk before I copy them into the allure placeholders.

Hopefully it works now :-)

I must admit though, that for what ever reason I couldnt get back to the state where the error happened on my virtual machine this evening. So I cant try the final code, which was refactored while I was writing the tests....

**Comment by jonn-set** [comment]
> Could you check if this solves the 256k and 1k issue?
> 

Sure will do. Thanks for taking time to fix the issues I report, appreciate it.

**Comment by jonn-set** [comment]
All looks good for with the current changes, videos larger than 256 KB generated and all files copied correctly into the `allure-raw` folder

Good to go from my side, thanks

**Comment by presidenten** [comment]
v2.0.0 is released and published on npm

**Comment by NavinSeth** [comment]
I just updated and the videos are not .mp4 for my Cucumber tests but instead .png. Am I missing something?

**Comment by presidenten** [comment]
> I just updated and the videos are not .mp4 for my Cucumber tests but instead .png. Am I missing something?

@NavinSeth 
Probably. Please download the cucumber branch of the template project and run `yarn && yarn e2e` and make sure that you get videos if you run `yarn report` afterwards.
https://github.com/presidenten/wdio-template/tree/cucumber

If it works, try comparing `wdio.conf.js` for differences with yours. 
If it doesnt work, please run the tests in the template with loglevel debug and send me the wdio-logs as gists.

**Comment by presidenten** [comment]
@NavinSeth Seems like I had some merge issues on the template project from the master branch. Just fixed it again. So if you forked or downloaded, please update before you try again.

**Comment by NavinSeth** [comment]
Can you publish a minor version with the new change?

**Comment by presidenten** [comment]
There is no change to the package, only the template projects cucumber branch:
https://github.com/presidenten/wdio-template/tree/cucumber

The demo directory in the package still works fine as well, just `yarn` in main dir, `yarn` in demo dir and `yarn e2e-cucumber`

**Comment by presidenten** [comment]
...the fix in wdio-template was actually to move head back to before the merge with the templates master branch :-)

**Pull request merged**